Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp

```diff
@@ -0,0 +1,348 @@
+                                                                LinOpBCType::bogus)});
+
+#ifdef WARPX_DIM_RZ
+    if (geom[0].ProbLo(0) == 0){
+        lobc[0] = LinOpBCType::Neumann;
+
+        // handle the r_max boundary explicitly
+        if (WarpX::field_boundary_hi[0] == FieldBoundaryType::PEC) {
+            hibc[0] = LinOpBCType::Dirichlet;
+        }
+        else if (WarpX::field_boundary_hi[0] == FieldBoundaryType::Neumann) {
+            hibc[0] = LinOpBCType::Neumann;
+        }
+        else {
+            WARPX_ALWAYS_ASSERT_WITH_MESSAGE(false,
+                "Field boundary condition at the outer radius must be either PEC or neumann "
+                "when using the Projection Divergence Cleaner"
+            );
+        }
+    }
+    const int dim_start = 1;
+#else
+    const int dim_start = 0;
+#endif
+    for (int idim=dim_start; idim<AMREX_SPACEDIM; idim++){
+        if ( WarpX::field_boundary_lo[idim] == FieldBoundaryType::Periodic
+             && WarpX::field_boundary_hi[idim] == FieldBoundaryType::Periodic ) {
+            lobc[idim] = LinOpBCType::Periodic;
+            hibc[idim] = LinOpBCType::Periodic;
+        }
+        else {
+            if ( WarpX::field_boundary_lo[idim] == FieldBoundaryType::PEC ) {
+                lobc[idim] = LinOpBCType::Dirichlet;
+            }
+            else if ( WarpX::field_boundary_lo[idim] == FieldBoundaryType::Neumann ) {
+                lobc[idim] = LinOpBCType::Neumann;
+            }
+            else {
+                WARPX_ABORT_WITH_MESSAGE(
+                    "Field boundary conditions have to be either periodic, PEC or neumann "
+                    "when using the Projection based Divergence Cleaner,  but they are " + GetFieldBCTypeString(WarpX::field_boundary_lo[idim])
+                );
+            }
+
+            if ( WarpX::field_boundary_hi[idim] == FieldBoundaryType::PEC ) {
+                hibc[idim] = LinOpBCType::Dirichlet;
+            }
+            else if ( WarpX::field_boundary_hi[idim] == FieldBoundaryType::Neumann ) {
+                hibc[idim] = LinOpBCType::Neumann;
+            }
+            else {
+                WARPX_ABORT_WITH_MESSAGE(
+                    "Field boundary conditions have to be either periodic, PEC or neumann "
+                    "when using the electrostatic Multigrid solver,  but they are " + GetFieldBCTypeString(WarpX::field_boundary_hi[idim])
+                );
+            }
+        }
+
+        WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+            (WarpX::field_boundary_lo[idim] != FieldBoundaryType::Open &&
+            WarpX::field_boundary_hi[idim] != FieldBoundaryType::Open &&
+            WarpX::field_boundary_lo[idim] != FieldBoundaryType::PML &&
+            WarpX::field_boundary_hi[idim] != FieldBoundaryType::PML) ,
+            "Open and PML field boundary conditions only work with "
+            "warpx.poisson_solver = fft."
```

Comment:
Currently, there is no check that avoids the div B cleaner from being executed if the fft solver is used, right? In that case this error message would be confusing, but it would be good to simply not perform the div B cleaning when the fft solver is used, and print a message warning about possible B-field divergence (as is done for the other cases).

Reply:
Doesn't the PSATD solver have its own flavor of div cleaning that diffuses it away over time? I guess since the BCs could be incompatible it should not be executed if using the PSATD. I will add another check for that.

Reply:
I tightened this check further to keep it from being executed with incompatible solvers.